hyperspy/axes.py

```diff
@@ -409,6 +357,203 @@ def _get_index(self, value):
+            text += ", index: %i" % self.index
+        text += ">"
+        return text
+
+    def __str__(self):
+        return self._get_name() + " axis"
+
+    def get_axis_dictionary(self):
+        return {'_type': self.__class__.__name__,
+                'name': _parse_axis_attribute(self.name),
+                'units': _parse_axis_attribute(self.units),
+                'navigate': self.navigate,
+                'is_binned': self.is_binned,
+                }
+
+    def copy(self):
+        return self.__class__(**self.get_axis_dictionary())
+
+    def __copy__(self):
+        return self.copy()
+
+    def __deepcopy__(self, memo):
+        cp = self.copy()
+        return cp
+
+    def update_from(self, axis, attributes):
+        """Copy values of specified axes fields from the passed AxesManager.
+
+        Parameters
+        ----------
+        axis : BaseDataAxis
+            The BaseDataAxis instance to use as a source for values.
+        attributes : iterable container of strings.
+            The name of the attribute to update. If the attribute does not
+            exist in either of the AxesManagers, an AttributeError will be
+            raised.
+
+        Returns
+        -------
+        A boolean indicating whether any changes were made.
+
+        """
+        any_changes = False
+        changed = {}
+        for f in attributes:
+            if getattr(self, f) != getattr(axis, f):
+                changed[f] = getattr(axis, f)
+        if len(changed) > 0:
+            self.trait_set(**changed)
+            any_changes = True
+        return any_changes
+
+
+class DataAxis(BaseDataAxis):
+    """DataAxis class for a non-uniform axis defined through an ``axis`` array.
+
+    The most flexible type of axis, where the axis points are directly given by
+    an array named ``axis``. As this can be any array, the property
+    ``is_uniform`` is automatically set to ``False``.
+
+    Parameters
+    ----------
+    axis : numpy array or list
+        The array defining the axis points.
+
+    Examples
+    --------
+    Sample dictionary for a `DataAxis`:
+
+    >>> dict0 = {'axis': np.arange(11)**2}
+    >>> s = hs.signals.Signal1D(np.ones(12), axes=[dict0])
+    >>> s.axes_manager[0].get_axis_dictionary()
+    {'_type': 'DataAxis',
+     'name': <undefined>,
+     'units': <undefined>,
+     'navigate': False,
+     'axis': array([  0,   1,   4,   9,  16,  25,  36,  49,  64,  81, 100])}
+    """
+
+    def __init__(self,
+                 index_in_array=None,
+                 name=None,
+                 units=None,
+                 navigate=False,
+                 is_binned=False,
+                 axis=[1],
+                 **kwargs):
+        super().__init__(
+            index_in_array=index_in_array,
+            name=name,
+            units=units,
+            navigate=navigate,
+            is_binned=is_binned,
+            **kwargs)
+        self.add_trait("_axis", t.Array)
```

Comment:
Why does it need to be a trait?

Reply:
I kept this just because I didn't want anyone trying to set the _axis with anything other than an array as that would complicate the fancy slicing quite a bit.

Reply:
I should note that the `self.events.any_axis_changed.trigger(obj=self)` should probably trigger when the `_axis` is changed.  I am not entirely sure what in the drawing is linked to that event.